File: vol1/ch13/LeetCode327.py
class BIT:                     #树状数组类
    def __init__(self,n):
        self.n=n
        self.tree=[0]*(n+1)					#树状数组为tree[1..n]
    def lowbit(self,x):
        return x&-x
    def UpdateOne(self,i,x):        #单点修改
        while i<=self.n:
            self.tree[i]+=x
            i+=self.lowbit(i)
    def Sum(self,i):            #求前缀和
        ans=0
        while i>0:
            ans+=self.tree[i]
            i-=self.lowbit(i)
        return ans
class Solution:
    def countRangeSum(self,nums:List[int],lower:int,upper:int)->int:
        n=len(nums)
        presum=[0]*(n+1)            #前缀和数组
        presum[0]=0
        for i in range(1,n+1):
            presum[i]=presum[i-1]+nums[i-1]
        a=[]                        #离散化
        for x in presum:
            a.append(x)
            a.append(x-lower)
            a.append(x-upper)
        hmap={}
        idx=0
        for x in sorted(a):
            if x not in hmap:
                hmap[x]=idx;idx+=1
        ans=0
        bit=BIT(len(hmap))
        for x in presum:
            i,j,k=hmap[x-upper],hmap[x-lower],hmap[x]
            ans+=bit.Sum(j+1)-bit.Sum(i)
            bit.UpdateOne(k+1,1)
        return ans

File: vol1/ch13/LeetCode327.cpp
#define lowbit(x) ((x)&-(x))
typedef long long LL;
class BIT {                     //树状数组类
private:
    vector<int> tree;					//树状数组为tree[1..n]
    int n;
public:
    BIT(int n):n(n),tree(n+1,0) {}
    void UpdateOne(int i,int x) {			//单点修改
        while (i<=n) {
            tree[i]+=x;
            i+=lowbit(i);
        }
    }
    int Sum(int i) {						//求前缀和
        int ans=0;
        while (i) {
            ans+=tree[i];
            i-=lowbit(i);
        }
        return ans;
    }
};
class Solution {
public:
    int countRangeSum(vector<int>& nums,int lower,int upper) {
        int n=nums.size();
        vector<LL> presum(n+1);         //前缀和数组
        presum[0]=0;
        for(int i=1;i<=n;i++)
            presum[i]=presum[i-1]+nums[i-1];
        set<LL> tset;                   				//除重
        for (LL x:presum) {
            tset.insert(x);
            tset.insert(x-lower);
            tset.insert(x-upper);
        }
        unordered_map<LL,int> hmap;     //离散化
        int idx=0;
        for(LL x:tset) {
            hmap[x]=idx++;
        }
        int ans=0;
        BIT bit(hmap.size());
        for(LL x: presum) {
            int i=hmap[x-upper],j=hmap[x-lower];
            ans+=bit.Sum(j+1)-bit.Sum(i);
            int k=hmap[x];
            bit.UpdateOne(k+1,1);
        }
        return ans;
    }
};

